armi/reactor/blueprints/blockBlueprint.py

```diff
@@ -115,11 +117,26 @@ def construct(
         self._checkByComponentMaterialInput(materialInput)
 
         for componentDesign in self:
-            filteredMaterialInput = self._filterMaterialInput(
+            filteredMaterialInput, byComponentMatModKeys = self._filterMaterialInput(
                 materialInput, componentDesign
             )
             c = componentDesign.construct(blueprint, filteredMaterialInput)
             components[c.name] = c
+
+            # check that the mat mods for this component are valid options
+            # this will only examine by-component mods, block mods are done later
+            if isinstance(c, Component):
```

Comment:
I'm trying to decide if `isinstance(c, Component)` is correct, if it is more correct (and future-proof) to use `hasattr(c, "material")`.

Why did you choose to only worry about Components that might have materials, and not, say, `Composite`s?

Just trying to find possible edge cases.

Reply:
Something that I stumbled on in the course of this work was the new `ComponentGroup`, which effectively sits at the same position in the reactor hierarchy as a `Component`:
https://github.com/terrapower/armi/blob/8b6ea1abcfe6ea4587ab0a2622f4d7c03e214e83/armi/reactor/blueprints/componentBlueprint.py#L160-L175

So when looping over all the "thing" in a block, you can get both `Component`s and `ComponentGroup`s (and maybe other things).

But it appeared to me that the only things that actually have `applyInputParam()` applied to them are `Component`s:
https://github.com/terrapower/armi/blob/4699c867e6b6fb7c1308a879e03c241f1d946b75/armi/reactor/blueprints/componentBlueprint.py#L238

So not knowing too much about these `ComponentGroup`s except that they don't receive material modifications, I just decided to test that the object was a `Component`.

But I guess your suggested approach would work as well.

After that explanation, do you want me to go in that direction? Happy to do so, just let me know.

Reply:
Ah!

I see the issue.  Sure, `ComponentGroup` is new and not fully implemented everywhere.  That's fine then. No worries.